File: sdks/typescript/src/processing.ts
/**
 * Processing Client
 */

import { HttpClient } from './http';
import {
  ProcessingJob,
  CreateJobRequest,
  PaginationParams,
  PaginatedResponse,
  JobStatus,
  ProcessingType,
} from './types';

export class ProcessingClient {
  constructor(private http: HttpClient) {}

  /**
   * Create a processing job
   */
  async createJob(request: CreateJobRequest): Promise<ProcessingJob> {
    const response = await this.http.post<any>('/processing/jobs', request);
    return this.transformJob(response);
  }

  /**
   * Get job by ID
   */
  async getJob(jobId: string): Promise<ProcessingJob> {
    const response = await this.http.get<any>(`/processing/jobs/${jobId}`);
    return this.transformJob(response);
  }

  /**
   * List jobs
   */
  async listJobs(
    params?: PaginationParams & {
      status?: JobStatus;
      type?: ProcessingType;
      fileId?: string;
    }
  ): Promise<PaginatedResponse<ProcessingJob>> {
    const response = await this.http.get<PaginatedResponse<any>>(
      '/processing/jobs',
      { params }
    );

    return {
      ...response,
      data: response.data.map(this.transformJob),
    };
  }

  /**
   * Cancel a job
   */
  async cancelJob(jobId: string): Promise<ProcessingJob> {
    const response = await this.http.delete<any>(`/processing/jobs/${jobId}`);
    return this.transformJob(response);
  }

  /**
   * Retry a failed job
   */
  async retryJob(jobId: string): Promise<ProcessingJob> {
    const response = await this.http.post<any>(`/processing/jobs/${jobId}/retry`);
    return this.transformJob(response);
  }

  /**
   * Wait for job completion
   */
  async waitForCompletion(
    jobId: string,
    options?: {
      pollInterval?: number;
      timeout?: number;
      onProgress?: (job: ProcessingJob) => void;
    }
  ): Promise<ProcessingJob> {
    const pollInterval = options?.pollInterval || 1000;
    const timeout = options?.timeout || 300000; // 5 minutes default
    const startTime = Date.now();

    while (true) {
      const job = await this.getJob(jobId);

      if (options?.onProgress) {
        options.onProgress(job);
      }

      if (job.status === 'completed' || job.status === 'failed' || job.status === 'cancelled') {
        return job;
      }

      if (Date.now() - startTime > timeout) {
        throw new Error(`Job ${jobId} did not complete within timeout`);
      }

      await new Promise((resolve) => setTimeout(resolve, pollInterval));
    }
  }

  // Convenience methods for common processing types

  /**
   * Resize an image
   */
  async resizeImage(
    fileId: string,
    options: {
      width?: number;
      height?: number;
      maintainAspectRatio?: boolean;
      format?: 'jpeg' | 'png' | 'webp';
      quality?: number;
    }
  ): Promise<ProcessingJob> {
    return this.createJob({
      fileId,
      type: 'image_resize',
      options,
    });
  }

  /**
   * Convert document format
   */
  async convertDocument(
    fileId: string,
    options: {
      targetFormat: 'pdf' | 'docx' | 'txt' | 'html';
      includeImages?: boolean;
      preserveFormatting?: boolean;
    }
  ): Promise<ProcessingJob> {
    return this.createJob({
      fileId,
      type: 'document_convert',
      options,
    });
  }

  /**
   * Transcode video
   */
  async transcodeVideo(
    fileId: string,
    options: {
      targetFormat: 'mp4' | 'webm' | 'avi';
      resolution?: '720p' | '1080p' | '4k';
      codec?: 'h264' | 'h265' | 'vp9';
      bitrate?: number;
    }
  ): Promise<ProcessingJob> {
    return this.createJob({
      fileId,
      type: 'video_transcode',
      options,
    });
  }

  /**
   * Scan file for viruses
   */
  async scanForViruses(fileId: string): Promise<ProcessingJob> {
    return this.createJob({
      fileId,
      type: 'virus_scan',
      priority: 'high',
    });
  }

  /**
   * Analyze file content
   */
  async analyzeContent(
    fileId: string,
    options?: {
      extractText?: boolean;
      detectLanguage?: boolean;
      extractEntities?: boolean;
      classifyContent?: boolean;
    }
  ): Promise<ProcessingJob> {
    return this.createJob({
      fileId,
      type: 'content_analysis',
      options,
    });
  }

  // Batch operations

  /**
   * Create batch processing jobs
   */
  async createBatchJobs(
    jobs: CreateJobRequest[]
  ): Promise<{ jobs: ProcessingJob[]; failed: Array<{ index: number; error: string }> }> {
    const response = await this.http.post<{
      jobs: any[];
      failed: Array<{ index: number; error: string }>;
    }>('/processing/jobs/batch', { jobs });

    return {
      jobs: response.jobs.map(this.transformJob),
      failed: response.failed,
    };
  }

  /**
   * Cancel multiple jobs
   */
  async cancelBatchJobs(jobIds: string[]): Promise<{
    cancelled: string[];
    failed: Array<{ jobId: string; error: string }>;
  }> {
    return this.http.post('/processing/jobs/batch/cancel', { jobIds });
  }

  /**
   * Get job statistics
   */
  async getStats(): Promise<{
    total: number;
    byStatus: Record<JobStatus, number>;
    byType: Record<ProcessingType, number>;
    averageProcessingTime: number;
  }> {
    return this.http.get('/processing/stats');
  }

  /**
   * Get job result
   */
  async getJobResult<T = unknown>(jobId: string): Promise<T> {
    const job = await this.getJob(jobId);

    if (job.status !== 'completed') {
      throw new Error(`Job ${jobId} is not completed (status: ${job.status})`);
    }

    return job.result as T;
  }

  /**
   * Get output file from job
   */
  async getOutputFile(jobId: string): Promise<string> {
    const result = await this.getJobResult<{ outputFileId: string }>(jobId);
    return result.outputFileId;
  }

  /**
   * Transform raw job response to ProcessingJob
   */
  private transformJob(raw: any): ProcessingJob {
    return {
      ...raw,
      createdAt: new Date(raw.createdAt),
      completedAt: raw.completedAt ? new Date(raw.completedAt) : undefined,
    };
  }
}

File: sdks/python/fileops/processing.py
"""
Processing Client
"""

from typing import Optional, Dict, Any
from datetime import datetime
import time
import httpx
from .types import ProcessingJob, ProcessingType, JobStatus, PaginatedResponse


class ProcessingClient:
    """File processing and job management client."""

    def __init__(self, http: httpx.Client):
        self._http = http

    def create_job(
        self,
        file_id: str,
        type: str,
        options: Optional[Dict[str, Any]] = None,
        priority: str = "normal",
        webhook_url: Optional[str] = None,
    ) -> ProcessingJob:
        """Create a processing job."""
        payload = {"fileId": file_id, "type": type, "priority": priority}
        if options:
            payload["options"] = options
        if webhook_url:
            payload["webhookUrl"] = webhook_url

        response = self._http.post("/processing/jobs", json=payload)
        response.raise_for_status()
        return self._parse_job(response.json())

    def get_job(self, job_id: str) -> ProcessingJob:
        """Get job by ID."""
        response = self._http.get(f"/processing/jobs/{job_id}")
        response.raise_for_status()
        return self._parse_job(response.json())

    def list_jobs(
        self,
        status: Optional[str] = None,
        type: Optional[str] = None,
        file_id: Optional[str] = None,
        page: int = 1,
        limit: int = 20,
    ) -> PaginatedResponse:
        """List processing jobs."""
        params = {"page": page, "limit": limit}
        if status:
            params["status"] = status
        if type:
            params["type"] = type
        if file_id:
            params["fileId"] = file_id

        response = self._http.get("/processing/jobs", params=params)
        response.raise_for_status()
        data = response.json()
        return PaginatedResponse(
            data=[self._parse_job(j) for j in data["data"]],
            page=data["pagination"]["page"],
            limit=data["pagination"]["limit"],
            total=data["pagination"]["total"],
            total_pages=data["pagination"]["totalPages"],
            has_next=data["pagination"]["hasNext"],
            has_prev=data["pagination"]["hasPrev"],
        )

    def cancel_job(self, job_id: str) -> ProcessingJob:
        """Cancel a job."""
        response = self._http.delete(f"/processing/jobs/{job_id}")
        response.raise_for_status()
        return self._parse_job(response.json())

    def wait_for_completion(
        self,
        job_id: str,
        poll_interval: float = 1.0,
        timeout: float = 300.0,
        on_progress: Optional[callable] = None,
    ) -> ProcessingJob:
        """Wait for job completion."""
        start_time = time.time()

        while True:
            job = self.get_job(job_id)

            if on_progress:
                on_progress(job)

            if job.status in ["completed", "failed", "cancelled"]:
                return job

            if time.time() - start_time > timeout:
                raise TimeoutError(f"Job {job_id} did not complete within timeout")

            time.sleep(poll_interval)

    # Convenience methods
    def resize_image(
        self,
        file_id: str,
        width: Optional[int] = None,
        height: Optional[int] = None,
        format: str = "jpeg",
        quality: int = 85,
    ) -> ProcessingJob:
        """Resize an image."""
        return self.create_job(
            file_id,
            "image_resize",
            {
                "width": width,
                "height": height,
                "format": format,
                "quality": quality,
            },
        )

    def convert_document(
        self, file_id: str, target_format: str = "pdf"
    ) -> ProcessingJob:
        """Convert document format."""
        return self.create_job(
            file_id, "document_convert", {"targetFormat": target_format}
        )

    def scan_for_viruses(self, file_id: str) -> ProcessingJob:
        """Scan file for viruses."""
        return self.create_job(file_id, "virus_scan", priority="high")

    def analyze_content(self, file_id: str) -> ProcessingJob:
        """Analyze file content."""
        return self.create_job(file_id, "content_analysis")

    def _parse_job(self, data: dict) -> ProcessingJob:
        return ProcessingJob(
            id=data["id"],
            type=data["type"],
            status=data["status"],
            file_id=data["fileId"],
            progress=data["progress"],
            result=data.get("result"),
            error=data.get("error"),
            created_at=datetime.fromisoformat(data["createdAt"].replace("Z", "+00:00"))
            if data.get("createdAt")
            else None,
            completed_at=datetime.fromisoformat(
                data["completedAt"].replace("Z", "+00:00")
            )
            if data.get("completedAt")
            else None,
        )
